### PR TITLE
Add sample bank selection controls (and hotkeys) to editor

### PR DESCRIPTION
**Pull request by peppy**
Mostly for the hotkeys, since they've come up multiple times.

https://user-images.githubusercontent.com/191335/196695476-fa20b4e7-c646-4114-aaa9-3aadf3f2e82c.mp4

- [x] Hotkeys need to be fixed to work post-merge of #20903

### CODE REVIEW

File: osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs [hunk 1]
```diff
@@ -104,12 +164,48 @@ protected virtual void UpdateTernaryStates()
...
+                h.SampleControlPoint.SampleBank = bankName;
+                EditorBeatmap.Update(h);
+            });
+        }
+
+        /// <summary>
+        /// Removes a sample bank from all selected <see cref="HitObject"/>s.
+        /// </summary>
+        /// <param name="bankName">The name of the sample bank.</param>
+        public void RemoveSampleBank(string bankName)
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs:198]: Method is not actually used. Not sure it makes any sense either?

File: osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs [hunk 2]
```diff
@@ -48,11 +49,70 @@ private void load()
...
         private void createStateBindables()
         {
+            foreach (string bankName in HitSampleInfo.AllBanks)
+            {
+                var bindable = new Bindable<TernaryState>
+                {
+                    Description = bankName.Titleize()
+                };
+
+                bindable.ValueChanged += state =>
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs:69]: Personal preference: This `ValueChanged` callback is a bit on the longer side. I'd probably split a method.

File: osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs [hunk 3]
```diff
@@ -186,6 +207,39 @@ protected virtual IEnumerable<TernaryButton> CreateTernaryButtons()
...
+                        Icon = FontAwesome.Solid.VolumeOff
+                    },
+                    new OsuSpriteText
+                    {
+                        Anchor = Anchor.CentreLeft,
+                        Origin = Anchor.CentreLeft,
+                        X = 10,
+                        Y = -1,
+                        Font = OsuFont.Default.With(weight: FontWeight.Bold, size: 20),
+                        Text = $"{char.ToUpper(sampleName.First())}"
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs:237]:
```suggestion
                        Text = $"{char.ToUpperInvariant(sampleName.First())}"
```

A little safer. Guess I didn't ban this one back when...

**Comment by peppy** [comment]
> Just a quick feedback, but it definely should be possible to select different banks for addtitions (whistle, finish, clap) and the hitobject itself, as combining those is a huge part of the hitsound stack and most of the hitsounds nowdays use it.

That has nothing to do with this change, right? People were asking for shortcut level bank switching and this adds that.

File: osu.Game/Rulesets/Edit/HitObjectComposer.cs [hunk 4]
```diff
@@ -146,6 +148,16 @@ private void load(OverlayColourProvider colourProvider)
                                         Direction = FillDirection.Vertical,
                                         Spacing = new Vector2(0, 5),
                                     },
+                                },
+                                new EditorToolboxGroup("bank (Shift-Q~R)")
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Rulesets/Edit/HitObjectComposer.cs:152]: Brings up a good point - this is misaligned due to no "auto".

**Comment by peppy** [comment]
Blocking this temporarily while I figure out what to do about "auto" (which doesn't really exist in lazer).

**Comment by peppy** [comment]
Note to self: `RadioButton` class exists and may be a better fit for this.

File: osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs]: I've removed this.

File: osu.Game/Rulesets/Edit/HitObjectComposer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Rulesets/Edit/HitObjectComposer.cs:166]: I've added back automatic selection and now things are realigned.

File: osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs:75]: Hmm, I dunno about this. I don't think splitting out is going to do much for readability given what the method is doing..

**Comment by peppy** [comment]
I've tidied this up (as much as it can be) and added test coverage.

I foresee this logic changing a bit once https://github.com/ppy/osu/pull/23445 is in, but I see both systems working together (just being more specific about where the banks are assigned to based on context). Failing that, we might be able to pull out this UI and use it in the popover as well (almost did that in this PR but feel we should get a base implementation in first).

**Comment by peppy** [comment]
The fail case here is where a `HitObject` has no samples, and the `All` condition here becomes `true` as a result:

JetBrains Rider 2023-05-25 at 08 51 40

This can only happen in tests, but I'll look to fix the logic regardless.

**Comment by bdach** [comment]
I've pushed in a few menial changes. Of particular note, https://github.com/ppy/osu/pull/20829/commits/7d8f08c0ea0e2e1f6d2952926f85557ff9447350 fixed a thankfully-non-issue which had `ComposeBlueprintContainer` briefly apply the `auto` bank to hitobjects. This was remedied because the automatic bank assignment feature runs essentially in `Update()`, so that `auto` bank was never user-facing, but I figure let's just avoid putting it there at all ever if it can be avoided.

There are a few wonky spots here, in particular I'm not sure how the automatic assignment feature is supposed to work with volume. The way it kinda works right now is that using automatic assignment will inherit not only the bank, but the volume as well, and if you switch to another bank then that volume will be kept - but if you, for instance:

1. choose auto bank
2. switch to soft
3. place an object

the placed object will inherit the volume of the previous object from the auto bank choice, but the new object pending placement after it will have 100 volume, which is probably not what is wanted. However, I'm not sure *what* is, and this feels like an important enough feature wherein I'm willing to look away, get this in, and see what mappers' expectations and feedback is going to be based on the way they'll interact with this. Or you could open another pull on top of this fixing that issue if you think you know a way to solve it. Either works for me.

**Review by bdach — APPROVED** [review]

**Pull request merged**